src/query/plan/operator.cpp

```diff
@@ -2450,17 +2458,18 @@ Filter::FilterCursor::FilterCursor(const Filter &self, utils::MemoryResource *me
       pattern_filter_cursors_(MakeCursorVector(self_.pattern_filters_, mem)) {}
 
 bool Filter::FilterCursor::Pull(Frame &frame, ExecutionContext &context) {
-  SCOPED_PROFILE_OP("Filter");
+  SCOPED_PROFILE_OP_BY_REF(self_);
 
   // Like all filters, newly set values should not affect filtering of old
   // nodes and edges.
   ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
                                 storage::View::OLD, context.frame_change_collector);
-  while (input_cursor_->Pull(frame, context)) {
+  while (input_cursor_->Pull(frame, context) || UNLIKELY(context.is_profile_query)) {
     for (const auto &pattern_filter_cursor : pattern_filter_cursors_) {
       pattern_filter_cursor->Pull(frame, context);
     }
     if (EvaluateFilter(evaluator, self_.expression_)) return true;
+    if (UNLIKELY(context.is_profile_query)) return false;
```

Comment:
This logic is a bit obscure to me. You are in the while loop if context.is_profile_query is true but then at the end you return false?

Reply:
The `for` loop executes in two cases:
* `input_cursor_->Pull` returns `true`
* in profile queries (even if there’s no input, the query plan should contain the branch(es) that do filtering)
This line force-exits the outer `while` loop for profile queries (in their case, the `while` condition is always true).

Reply:
Okay but if the context.is_profile_query is true what do you need to execute, for loop of filter evaluation? Because it is a bit not clear that you want to got to while loop just to execute it once. IMO it is better to extract the part you need, not sure if it is for loop or evaluate filter, something like this:

```
if (context.is_profile_query) {
  for loop and/or
  evaluate filter
} else {
  while (input_cursor->Pull...) ...

```